### PR TITLE
[BCI-1372] In Memory Head Saver

### PR DESCRIPTION
**Pull request by ExtremelySunnyYK**
This PR is a POC implementation of the In Memory HeadSaver. This PR **should not to be merged to develop**, but merely for the reviews!

#### High Level Summary

The In-Memory HeadSaver project is a proposed solution to the current challenges faced by the HeadTracker Generalisation system when dealing with Non-EVM chains. 

The key goal is to create an In-Memory HeadSaver for Solana that stores 'Head' data internally without the need for long-term database storage. The proposed In-Memory HeadSaver is designed to be an internal feature of the HeadSaver package, remaining isolated from other chain-specific components for simplicity and abstraction.

The main approach involves using a Dictionary Only setup, storing heads in memory using a dictionary data structure, and setting the linked list relationship between each head when added. The dictionary provides time efficiency and uniqueness, but does not preserve order. To manage head numbers, an additional map data structure is proposed.

### Testing

https://github.com/smartcontractkit/chainlink/assets/46377366/41f00760-1b61-44e0-ac83-84a74d9c85b5

I have included a script for testing along with the video above. 

Run `./in-memory-test.sh` to run the tests.

### CODE REVIEW

File: common/types/chain.go [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package types
+
+import "fmt"
+
+// TODO : Remove ID from txmgr/types to move to common/types
```
Comment (prashantkumar1982) [inline common/types/chain.go:5]: Ideally I would also like to merge this with https://github.com/smartcontractkit/chainlink/blob/d5b44c27abd138ee6be4be02f2bc243a52f03e1a/core/chains/constraints.go#L4.

Perhaps @jmank88 can weigh in.

Comment (jmank88) [reply common/types/chain.go:5]: sure if they are similar enough

File: core/chains/evm/headtracker/head_saver_in_mem.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+		evmConfig,
+		lggr,
+		func() *evmtypes.Head { return nil },
+		func(head, parent *evmtypes.Head) { head.Parent = parent },
+	)
+}
+
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) Save(ctx context.Context, head H) error {
+	if !head.IsValid() {
+		return nil
```
Comment (prashantkumar1982) [inline core/chains/evm/headtracker/head_saver_in_mem.go:62]: should be return error here

File: core/chains/evm/headtracker/head_saver_in_mem.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	defer hs.mu.Unlock()
+
+	hs.trimHeads(historyDepth)
+}
+
+// trimHeads() is should only be called by functions with mutex locking.
+// trimHeads() is an internal function without locking to prevent deadlocks
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) trimHeads(historyDepth int64) {
+	for headNumber, headNumberList := range hs.HeadsNumber {
+		if headNumber < historyDepth {
```
Comment (prashantkumar1982) [inline core/chains/evm/headtracker/head_saver_in_mem.go:148]: If historyDepth is just a number indicating the depth to keep, then this needs to be:
if hs.latestHead.Number - headNumber > historyDepth

Comment (prashantkumar1982) [reply core/chains/evm/headtracker/head_saver_in_mem.go:148]: The code here assumes historyDepth = the oldest head number that we need to keep

**Review by prashantkumar1982 — COMMENTED** [review]
Overall, looks very good. Nice job!

File: core/chains/evm/headtracker/head_saver_in_mem.go (hunk 2 above)
Comment (ExtremelySunnyYK) [reply core/chains/evm/headtracker/head_saver_in_mem.go:62]: thanks for pointing this out, my mistake for not handling the error here.

**Pull request closed without merging**